### PR TITLE
Add an `auth` module that can help users generate tokens 

### LINKED ISSUE #607: Support getting an authentication token from within Hera
**Is your feature request related to a problem? Please describe.**

Currently, to authenticate with the Argo server, a token has to be provided by an outside process (e.g. `argo auth token`). This is frustrating to work with when trying to automate Workflow submission and management with Hera. 

**Describe the solution you'd like**
Provide a direct interface in Hera that allows an already authenticated user to access the authentication token with a built-in API, e.g. `

```
from hera import Auth
from hera.workflows import WorkflowsService

token = Auth.get_token()
svc = WorkflowsService(host=argo_uri, token=token)
```

**Describe alternatives you've considered**

One can run `os.subprocess` and capture the output from `argo auth token`, but this is cumbersome and has known security gotchas if used improperly (although those may not apply as much in this example). 

**Additional context**

Brief Slack discussion [here](https://cloud-native.slack.com/archives/C03NRMD9KPY/p1682968103747879)

### PR DESCRIPTION
**Pull request by flaviuvadan**
**Pull Request Checklist**
- [x] Fixes #607 
- [ ] Tests added
- [x] Documentation/examples added
- [x] [Good commit messages](https://cbea.ms/git-commit/) and/or PR title

**Description of PR**

Currently, users have to write their own token generation, even if it uses something like the Argo token or K8s SA tokens. 

This PR adds an `auth` module that can host common token generators.

### CODE REVIEW

File: src/hera/shared/_global_config.py [hunk 1]
```diff
@@ -28,17 +30,17 @@ class _GlobalConfig:
 
     # protected attributes are ones that are computed/go through some light processing upon setting or
     # are processed upon accessing. The rest, which use primitive types, such as `str`, can remain public
-    _token: Union[Optional[str], Callable[[], Optional[str]]] = None
+    _token: Union[Optional[str], TokenGenerator, Callable[[], Optional[str]]] = None
```
Comment (elliotgunton) [inline src/hera/shared/_global_config.py:33]: Should the `Optional` be on the outside?
```suggestion
    _token: Optional[Union[str], TokenGenerator, Callable[[], Optional[str]]] = None
```

File: src/hera/shared/_global_config.py [hunk 2]
```diff
@@ -28,17 +30,17 @@ class _GlobalConfig:
 
     # protected attributes are ones that are computed/go through some light processing upon setting or
     # are processed upon accessing. The rest, which use primitive types, such as `str`, can remain public
-    _token: Union[Optional[str], Callable[[], Optional[str]]] = None
+    _token: Optional[Union[str], TokenGenerator, Callable[[], Optional[str]]] = None
```
Comment (elliotgunton) [inline src/hera/shared/_global_config.py:33]: I messed up the brackets sorry!
```suggestion
    _token: Optional[Union[str, TokenGenerator, Callable[[], Optional[str]]]] = None
```

**Review by elliotgunton — APPROVED** [review]
LGTM!

**Comment by samj1912** [comment]
@flaviuvadan can we look at an alternative to the k8s Auth module for generating tokens? Would it be easier to just copy/vendor it?

**Comment by flaviuvadan** [comment]
> @flaviuvadan can we look at an alternative to the k8s Auth module for generating tokens? Would it be easier to just copy/vendor it?

Can you say more about this? I am not sure what you mean by copy/vendor it. Do you mean use a K8s library for this type of token generation rather than maintain this code in Hera?

**Review by samj1912 — APPROVED** [review]

**Pull request merged**